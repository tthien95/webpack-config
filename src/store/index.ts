import { configureStore } from '@reduxjs/toolkit';
import toastSlice from './toast-slice';

const store = configureStore({
  reducer: toastSlice,
});

export default store;
