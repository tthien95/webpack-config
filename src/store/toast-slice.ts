/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type Notification = {
  status: 'success' | 'error';
  title: string;
  message: string;
};

export interface ToastState {
  notification: Notification | null;
}

const initialState: ToastState = { notification: null };

const toastSlice = createSlice({
  name: 'toast',
  initialState,
  reducers: {
    showNotification: (state, action: PayloadAction<Notification>) => {
      const { status, title, message } = action.payload;
      state.notification = { status, title, message };
    },
    hideNotification: (state) => {
      state.notification = null;
    },
  },
});

export const toastActions = toastSlice.actions;

export default toastSlice.reducer;
