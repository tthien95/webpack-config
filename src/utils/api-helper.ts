import axios from 'axios';

export const baseUrl = 'https://dummyjson.com';

export const get = (url: string) => axios.get(baseUrl + url);

export const post = (url: string, body: unknown) =>
  axios.post(baseUrl + url, body);

export const put = (url: string, body: unknown) =>
  axios.put(baseUrl + url, body);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const deleteReq = (url: string, body: any = {}) =>
  axios.delete(baseUrl + url, body);
