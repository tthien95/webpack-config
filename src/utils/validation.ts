export interface ValRes {
  isValid: boolean;
  message: string;
}
export interface Validations {
  [validatedField: string]: {
    required: boolean;
    validator: string | null;
  };
}

const validators: {
  [validateFunc: string]: (value: string) => ValRes;
} = {
  validateEmail: (value: string) => ({
    isValid:
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(
        value
      ),
    message: 'Invalid email format',
  }),

  validatePhone: (value: string) => ({
    isValid:
      /^\+?([0-9]{2})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/.test(
        value
      ),
    message:
      'Invalid phone number. Must contain + followed by country code and 10 digits phone number',
  }),
  validateNoSpecialChar: (value: string) => ({
    isValid: /^[A-Za-z]*$/.test(value),
    message: 'Must contains only alphabet characters',
  }),
};

export default validators;
