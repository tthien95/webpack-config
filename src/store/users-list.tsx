/* eslint-disable @typescript-eslint/no-empty-function */
import React, { useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { AxiosError } from 'axios';
import { toastActions } from './toast-slice';
import { User, UsersListContextType } from '../type/user';

const UsersListContext = React.createContext<UsersListContextType>({
  usersList: [],
  setUsersList: () => {},
  isLoading: false,
  addUser: () => {},
  updateUser: () => {},
  deleteUser: () => {},
  setIsLoading: () => {},
  fnHandleError: () => {},
});

export const UserContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const [usersList, setUsersList] = useState<User[] | []>([]);
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useDispatch();

  const fnHandleError = useCallback(
    ({ response }: AxiosError<{ message: string }>) => {
      dispatch(
        toastActions.showNotification({
          status: 'error',
          title: 'Error',
          message:
            response?.data.message ||
            response?.statusText ||
            'There is something wrong happended while fetching data',
        })
      );
    },
    [dispatch]
  );

  const addUser = (value: User) => {
    const currUsersList = [...usersList];
    currUsersList.unshift(value);
    setUsersList(currUsersList);
  };

  const updateUser = (value: User) => {
    const currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == value.id);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1, value);
      setUsersList(currUsersList);
    }
  };

  const deleteUser = (userId: string) => {
    const currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == userId);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1);
      setUsersList(currUsersList);
    }
  };

  return (
    <UsersListContext.Provider
      value={{
        usersList,
        setUsersList,
        addUser,
        updateUser,
        deleteUser,
        isLoading,
        setIsLoading,
        fnHandleError,
      }}
    >
      {children}
    </UsersListContext.Provider>
  );
};

export default UsersListContext;
