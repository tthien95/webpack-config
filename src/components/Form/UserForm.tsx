import { useCallback, useContext } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { AxiosResponse } from 'axios';
import { FormFields, FieldSet } from 'type/form';
import { User } from 'type/user';
import { post, put } from 'utils/api-helper';
import { Validations } from 'utils/validation';
import useForm from 'hooks/use-form';
import UsersListContext from 'store/users-list';
import { toastActions } from 'store/toast-slice';
import FormFieldSet from './FormFieldSet';
import Spinner from '../UI/Spinner';

const initialValues: FormFields = {
  firstName: '',
  lastName: '',
  birthDate: '',
  email: '',
  phone: '',
};

const validations: Validations = {
  firstName: {
    required: true,
    validator: 'validateNoSpecialChar',
  },
  lastName: {
    required: true,
    validator: 'validateNoSpecialChar',
  },
  birthDate: {
    required: true,
    validator: null,
  },
  email: {
    required: true,
    validator: 'validateEmail',
  },
  phone: {
    required: false,
    validator: 'validatePhone',
  },
};

const fieldSet: FieldSet[] = [
  {
    fieldSetLabel: 'Personal Info: ',
    fields: [
      {
        inputName: 'firstName',
        inputLabel: 'First Name',
        inputType: 'text',
      },
      {
        inputName: 'lastName',
        inputLabel: 'Last Name',
        inputType: 'text',
      },
      {
        inputName: 'birthDate',
        inputLabel: 'Birth Date',
        inputType: 'date',
      },
    ],
  },
  {
    fieldSetLabel: 'Contact Info: ',
    fields: [
      {
        inputName: 'email',
        inputLabel: 'Email address',
        inputType: 'email',
      },
      {
        inputName: 'phone',
        inputLabel: 'Phone',
        inputType: 'tel',
      },
    ],
  },
];

const UserForm = () => {
  const { userId } = useParams();
  const { updateUser, addUser, fnHandleError } = useContext(UsersListContext);
  const dispatch = useDispatch();

  const navigate = useNavigate();

  const fnSuccess = useCallback(
    (
      res: AxiosResponse<User>,
      ctxHandler: (e: User) => void,
      message: string
    ) => {
      const { id, image, email, birthDate, phone, firstName, lastName } =
        res.data;
      ctxHandler({ id, image, email, birthDate, phone, firstName, lastName });
      dispatch(
        toastActions.showNotification({
          status: 'success',
          title: 'Success',
          message,
        })
      );
      navigate('/');
    },
    [dispatch, navigate]
  );

  const onSubmit = async () => {
    let url = '/users/add';
    if (userId) {
      url = `/users/${userId}`;

      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      return put(url, data).then((res) => {
        fnSuccess(res, updateUser, 'Update request has been sent successfully');
      }, fnHandleError);
    }

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return post(url, data).then((res) => {
      fnSuccess(res, addUser, 'Add request has been sent successfully');
    }, fnHandleError);
  };

  const { data, handleChange, handleSubmit, loading, error } = useForm({
    initialValues,
    validations,
    userId,
    onSubmit,
  });

  if (loading) {
    return <Spinner />;
  }

  return (
    <form className="container-md mt-3" onSubmit={handleSubmit}>
      {fieldSet.map((set, index) => (
        <FormFieldSet
          {...set}
          key={index}
          inputValues={data}
          errorMess={error}
          handleChange={handleChange}
        />
      ))}
      <div className="row container">
        <div className="col-6">
          <button type="submit" className="btn btn-primary">
            Submit
          </button>
        </div>
      </div>
    </form>
  );
};

export default UserForm;
