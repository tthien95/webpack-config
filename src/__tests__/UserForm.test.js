import {
  cleanup,
  render,
  waitForElementToBeRemoved,
  screen,
  waitFor
} from '@testing-library/react';
import UserForm from '../components/Form/UserForm';

import { useSelector, useDispatch, Provider } from 'react-redux';
import store from '../store/index';
import { act, Simulate } from 'react-dom/test-utils';
import { toastActions } from '../store/toast-slice';
import axios from 'axios';
import { useNavigate, useParams } from 'react-router-dom';
import UsersListContext from '../store/users-list';
import { baseUrl } from '../utils/api-helper';
import userEvent from '@testing-library/user-event';

const sampleUser = {
  id: 1,
  firstName: 'Terry',
  lastName: 'Medhurst',
  birthDate: '2000-12-25',
  email: 'atuny0@sohu.com',
  phone: '+63 791 675 8914'
};

const fields = {
  firstName: 'First Name',
  lastName: 'Last Name',
  birthDate: 'Birth Date',
  email: 'Email address',
  phone: 'Phone'
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
  useParams: jest.fn()
}));

jest.mock('axios');

const wrapperRender = ({
  updateUser = () => {},
  addUser = () => {},
  fnHandleError = () => {}
} = {}) => {
  return render(
    <Provider store={store}>
      <UsersListContext.Provider value={{ updateUser, addUser, fnHandleError }}>
        <UserForm />
      </UsersListContext.Provider>
    </Provider>
  );
};

describe('UserForm', () => {
  afterAll(cleanup);

  beforeEach(() => {
    useDispatch.mockReturnValue(() => {});
    useNavigate.mockReturnValue(() => {});
    useSelector.mockReturnValue(() => {});
  });

  it('should display empty form for new user', () => {
    useParams.mockReturnValue({ userId: null });

    const { container } = wrapperRender();

    expect(container).toMatchSnapshot();
  });

  it('should display form with user information for edit', async () => {
    useParams.mockReturnValue({ userId: '1' });
    axios.get.mockResolvedValue({
      data: { ...sampleUser }
    });

    const { container } = wrapperRender();
    await waitForElementToBeRemoved(() => screen.queryByRole('status'));
    expect(axios.get).toBeCalled();
    expect(axios.get).toBeCalledWith(`${baseUrl}/users/${sampleUser.id}`);

    expect(container).toMatchSnapshot();
  });

  it('should submit add user form with correct inputs', async () => {
    const testData = {
      firstName: 'abc',
      lastName: 'def',
      birthDate: '2000-12-25',
      email: 'abc@def.com',
      phone: '+63 791 675 8914'
    };
    useParams.mockReturnValue({ userId: null });

    const dispatch = jest.fn();

    const navigate = jest.fn();
    const addUser = jest.fn();

    useDispatch.mockReturnValue(dispatch);
    useNavigate.mockReturnValue(navigate);
    axios.post.mockResolvedValue({
      data: {
        id: 1,
        ...testData
      }
    });

    wrapperRender({ addUser });

    for (const key in fields) {
      if (Object.hasOwnProperty.call(fields, key)) {
        const element = fields[key];
        userEvent.type(screen.getByLabelText(element), testData[key]);
      }
    }

    userEvent.click(screen.getByRole('button'));

    expect(axios.post).toBeCalled();
    expect(axios.post).toBeCalledWith(`${baseUrl}/users/add`, testData);
    await waitFor(() => expect(addUser).toBeCalled());
    expect(dispatch).toBeCalled();
    expect(dispatch).toBeCalledWith(
      toastActions.showNotification({
        status: 'success',
        title: 'Success',
        message: 'Add request has been sent successfully'
      })
    );
    expect(navigate).toBeCalled();
    expect(navigate).toBeCalledWith('/');
  });

  it('should submit edit request when inputs are all correct', async () => {
    useParams.mockReturnValue({ userId: '1' });

    const newInput = {
      ...sampleUser,
      firstName: 'Ter',
      lastName: 'Med'
    };

    const dispatch = jest.fn();

    const navigate = jest.fn();
    const updateUser = jest.fn();

    axios.get.mockResolvedValue({
      data: { ...sampleUser }
    });

    axios.put.mockResolvedValue({
      data: { ...newInput }
    });

    useDispatch.mockReturnValue(dispatch);
    useNavigate.mockReturnValue(navigate);

    delete newInput.id;

    wrapperRender({ updateUser });
    await waitForElementToBeRemoved(() => screen.queryByRole('status'));
    expect(axios.get).toBeCalled();
    expect(axios.get).toBeCalledWith(`${baseUrl}/users/${sampleUser.id}`);

    act(() => {
      Simulate.change(screen.getByLabelText('First Name'), {
        target: { value: newInput.firstName, name: 'firstName' }
      });

      Simulate.change(screen.getByLabelText('Last Name'), {
        target: { value: newInput.lastName, name: 'lastName' }
      });
    });

    userEvent.click(screen.getByRole('button'));

    expect(axios.put).toBeCalled();
    expect(axios.put).toBeCalledWith(
      `${baseUrl}/users/${sampleUser.id}`,
      newInput
    );
    await waitFor(() => expect(updateUser).toBeCalled());
    expect(dispatch).toBeCalled();
    expect(dispatch).toBeCalledWith(
      toastActions.showNotification({
        status: 'success',
        title: 'Success',
        message: 'Update request has been sent successfully'
      })
    );
    expect(navigate).toBeCalled();
    expect(navigate).toBeCalledWith('/');
  });

  it('should show error messages for wrong inputs', () => {
    const testData = {
      firstName: '',
      lastName: '',
      birthDate: '',
      email: 'abc',
      phone: '638914'
    };

    useParams.mockReturnValue({ userId: null });

    const { container } = wrapperRender();

    act(() => {
      for (const key in fields) {
        if (Object.hasOwnProperty.call(fields, key)) {
          const element = fields[key];
          Simulate.change(screen.getByLabelText(element), {
            target: { value: testData[key], name: key }
          });
        }
      }
    });

    userEvent.click(screen.getByRole('button'));

    expect(container).toMatchSnapshot();
    expect(axios.post).not.toBeCalled();
  });

  it('should show error and redirect to home', async () => {
    useParams.mockReturnValue({ userId: '11241515' });

    axios.get.mockRejectedValue({
      response: {
        statusText: 'Error message'
      }
    });

    const dispatch = jest.fn();
    const navigate = jest.fn();

    useDispatch.mockReturnValue(dispatch);
    useNavigate.mockReturnValue(navigate);

    wrapperRender();
    await waitForElementToBeRemoved(() => screen.queryByRole('status'));
    expect(axios.get).toBeCalled();
    expect(axios.get).toBeCalledWith(`${baseUrl}/users/11241515`);

    expect(dispatch).toBeCalled();
    expect(dispatch).toBeCalledWith(
      toastActions.showNotification({
        status: 'error',
        title: 'Error',
        message: 'Error message'
      })
    );
    expect(navigate).toBeCalled();
    expect(navigate).toBeCalledWith('/');
  });

  it('should handle error in success callback', async () => {
    useParams.mockReturnValue({ userId: '1' });
    axios.get.mockResolvedValue({
      message: 'Error message'
    });

    const dispatch = jest.fn();
    const navigate = jest.fn();

    useDispatch.mockReturnValue(dispatch);
    useNavigate.mockReturnValue(navigate);

    wrapperRender();
    await waitForElementToBeRemoved(() => screen.queryByRole('status'));
    expect(axios.get).toBeCalled();
    expect(axios.get).toBeCalledWith(`${baseUrl}/users/1`);

    expect(dispatch).toBeCalled();
    expect(dispatch).toBeCalledWith(
      toastActions.showNotification({
        status: 'error',
        title: 'Error',
        message: 'Error message'
      })
    );
    expect(navigate).toBeCalled();
    expect(navigate).toBeCalledWith('/');
  });
});
