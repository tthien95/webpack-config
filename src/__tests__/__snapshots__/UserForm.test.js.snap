// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm should display empty form for new user 1`] = `
<div>
  <form
    class="container-md mt-3"
  >
    <fieldset
      class="row container"
    >
      <legend>
        Personal Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="birthDate"
          >
            Birth Date
          </label>
          <input
            class="form-control"
            id="birthDate"
            name="birthDate"
            type="date"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <fieldset
      class="row container"
    >
      <legend>
        Contact Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="email"
          >
            Email address
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="phone"
          >
            Phone
          </label>
          <input
            class="form-control"
            id="phone"
            name="phone"
            type="tel"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row container"
    >
      <div
        class="col-6"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`UserForm should display form with user information for edit 1`] = `
<div>
  <form
    class="container-md mt-3"
  >
    <fieldset
      class="row container"
    >
      <legend>
        Personal Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            type="text"
            value="Terry"
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            type="text"
            value="Medhurst"
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="birthDate"
          >
            Birth Date
          </label>
          <input
            class="form-control"
            id="birthDate"
            name="birthDate"
            type="date"
            value="2000-12-25"
          />
        </div>
      </div>
    </fieldset>
    <fieldset
      class="row container"
    >
      <legend>
        Contact Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="email"
          >
            Email address
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            type="email"
            value="atuny0@sohu.com"
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="phone"
          >
            Phone
          </label>
          <input
            class="form-control"
            id="phone"
            name="phone"
            type="tel"
            value="+63 791 675 8914"
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row container"
    >
      <div
        class="col-6"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`UserForm should show error messages for wrong inputs 1`] = `
<div>
  <form
    class="container-md mt-3"
  >
    <fieldset
      class="row container"
    >
      <legend>
        Personal Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="firstName"
          >
            First Name
          </label>
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="birthDate"
          >
            Birth Date
          </label>
          <input
            class="form-control"
            id="birthDate"
            name="birthDate"
            type="date"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <fieldset
      class="row container"
    >
      <legend>
        Contact Info: 
      </legend>
      <div
        class="row mb-3"
      >
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="email"
          >
            Email address
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            type="email"
            value="abc"
          />
        </div>
        <div
          class="col-6"
        >
          <label
            class="form-label"
            for="phone"
          >
            Phone
          </label>
          <input
            class="form-control"
            id="phone"
            name="phone"
            type="tel"
            value="638914"
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row container"
    >
      <div
        class="col-6"
      >
        <button
          class="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
